File: NCoreUtils.AspNetCore.Rest/DefaultDefaultOrderProperty.fs
namespace NCoreUtils.AspNetCore.Rest

open System
open System.Reflection

[<AutoOpen>]
module private DefaultDefaultOrderPropertyHelpers =

  let inline eq a b = StringComparer.OrdinalIgnoreCase.Equals (a, b)

  let inline name (propertyInfo : PropertyInfo) = propertyInfo.Name

  let inline eqname value propertyInfo = name propertyInfo |> eq value

  let inline (|HasNamedProperty|_|) name (properties : PropertyInfo[]) =
    properties |> Array.tryFind (eqname name)


/// Provides default implementation for retrieving default ordering property.
type DefaultDefaultOrderProperty () =
  abstract Select<'a> : unit -> OrderByProperty
  /// <summary>
  /// Default implementation for retrieving default ordering property. Probes properties in following order: Updated,
  /// Created, Id, first defined property.
  /// </summary>
  /// <typeparam name="a">Type of the object ot return default ordering property for.</typeparam>
  /// <returns>
  /// Tuple containing default ordering property and whether default ordering direction is descending.
  /// </returns>
  default __.Select<'a> () =
      match typeof<'a>.GetProperties(BindingFlags.Instance ||| BindingFlags.Public) with
      | HasNamedProperty "Updated" p -> { Property = p; IsDescending = true }
      | HasNamedProperty "Created" p -> { Property = p; IsDescending = true }
      | HasNamedProperty "Id"      p -> { Property = p; IsDescending = false }
      | properties                   -> { Property = properties |> Array.head; IsDescending = false }

  interface IDefaultOrderProperty with
    member this.Select<'a> () = this.Select<'a> ()

  /// Gets shared instance of the default implementation for retrieving default ordering property.
  static member val SharedInstance = DefaultDefaultOrderProperty ()

File: NCoreUtils.AspNetCore.Rest/Rest/DefaultDefaultOrderProperty.cs
using System;
using System.Collections.Concurrent;
using System.Diagnostics.CodeAnalysis;
using System.Linq;
using System.Reflection;

namespace NCoreUtils.AspNetCore.Rest
{
    public class DefaultDefaultOrderProperty
    {
        static ConcurrentDictionary<Type, OrderByProperty> _cache = new ConcurrentDictionary<Type, OrderByProperty>();

        static Func<Type, OrderByProperty> _factory = Create;

        [UnconditionalSuppressMessage("Trimming", "IL2070", Justification = "Ensured by caller.")]
        private static OrderByProperty Create(Type type)
        {
            var properties = type.GetProperties(BindingFlags.Instance | BindingFlags.Public);
            PropertyInfo? updatedProperty = null;
            PropertyInfo? createdProperty = null;
            PropertyInfo? idProperty = null;
            foreach (var property in properties)
            {
                if (StringComparer.OrdinalIgnoreCase.Equals("Updated", property.Name))
                {
                    updatedProperty = property;
                }
                else if (StringComparer.OrdinalIgnoreCase.Equals("Created", property.Name))
                {
                    createdProperty = property;
                }
                else if (StringComparer.OrdinalIgnoreCase.Equals("Id", property.Name))
                {
                    idProperty = property;
                }
            }
            if (null != updatedProperty)
            {
                return new OrderByProperty(updatedProperty, true);
            }
            if (null != createdProperty)
            {
                return new OrderByProperty(createdProperty, true);
            }
            if (null != idProperty)
            {
                return new OrderByProperty(idProperty, false);
            }
            return new OrderByProperty(properties.First(), false);
        }

        protected DefaultDefaultOrderProperty() { }

        protected internal static OrderByProperty GetDefaultOrderByProperty([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)] Type type)
            => _cache.GetOrAdd(type, _factory);
    }

    public class DefaultDefaultOrderProperty<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)] T>
        : DefaultDefaultOrderProperty, IDefaultOrderProperty<T>
    {
        public OrderByProperty Select() => GetDefaultOrderByProperty(typeof(T));
    }
}